[Security Solution][feat] Alert filters for Detection Page

logeekal opened the pull request:
## Summary

This PR adds feature according to https://github.com/elastic/security-team/issues/4128.

### Capabilities Added

- 💚 Page Filters are dynamic component that have been added on Detection Alerts Page. 
- 🌀 The status of the page filters is propagated to the URL and any external page can direct users to Detection Alert Page with custom filters populated.
- 🌴 Filters are `hierarchical` in nature, what it means is that they should be arranged from lowers cardinality to highest from left to right. Any change in the filter on the left side, will update all the filters on the right hand side.
- 💊 Existing alerts status change pill ( open/acknowledged/closed) has been removed from the alerts page but ❗it still exists on the rule details page. 
- 📄 Filters are persisted during user's browser session through local storage but they are not shared across spaces.

📔  Here is how the page looks when alert index exists v/s if there is some error in fetching the index.

| Working Page Filters | When Index is unavailable |
|--|--|
| <video src="https://user-images.githubusercontent.com/7485038/208958595-16eddb1b-cef2-4598-afc2-a69d36ee1d1a.mov" />|![Screenshot 2022-12-06 at 13 48 57](https://user-images.githubusercontent.com/7485038/205947352-512b75d2-0c78-40a0-b713-b1282efc165f.png)|






### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

logeekal committed:
feat: detection page filter initial commit

logeekal committed:
refactor: add cypress tests & exclude+exists support

logeekal committed:
fix: issue with rule details page

elasticmachine commented:
Pinging @elastic/kibana-presentation (Team:Presentation)

spalger reviewed (approved):
Operations changes LGTM, though I'm curious how much of an increase this caused to the page load asset size and why.

logeekal committed:
merging main in current branch

logeekal committed:
Merging main in

logeekal committed:
fixing merge issues

logeekal committed:
feat: add reset popover + reset to default

logeekal committed:
remove unncessary files

logeekal committed:
Merging main ➡️  current branch

logeekal committed:
Test incremental commit

michaelolo24 commented on x-pack/plugins/security_solution/public/common/hooks/use_url_state.ts:
```diff
@@ -30,5 +30,6 @@ export enum URL_PARAM_KEY {
   sourcerer = 'sourcerer',
   timeline = 'timeline',
   timerange = 'timerange',
+  pageFilter = 'pageFilters',
```
I wonder if we should name this something more specific to the alerts page or keep this generic in case we want to use these filters on another page (i.e. events?).. Just thinking aloud

logeekal replied:
@michaelolo24 Actually, this param key is tied to to the filter group component which can be reused across other pages as well. May be we can rename it as `filterGroup` or anything else. 

Additionaly, I can take param key as the prop the component so that implementer can choose it. what do you think?

michaelolo24 replied:
Nah, no reason to, better to remain consistent across the pages. Thanks!

michaelolo24 commented on x-pack/plugins/security_solution/public/common/components/filter_group/index.tsx:
```diff
@@ -0,0 +1,315 @@
+          </EuiPopover>
+        </EuiFlexItem>
+        {/*
+         *<EuiFlexItem grow={false}>
```
Figure this is adding functionality for users to manually add filters on the page. Not sure if you want to delete this, or add a comment explaining why it's currently commented out?

logeekal replied:
@michaelolo24 

Yes, that was the initial intention so that the component is composable and easily understandable but current controls plugin do no support this way to defining components so I had to change the strategy.

I will remove it. Thanks.

michaelolo24 commented on x-pack/plugins/security_solution/public/common/components/filter_group/index.tsx:
```diff
@@ -0,0 +1,315 @@
+    }
+  };
+
+  useInitializeUrlParam(URL_PARAM_KEY.pageFilter, onUrlParamInit);
```
I would add a comment to the `URL_PARAM_KEY` in `x-pack/plugins/security_solution/public/common/hooks/use_url_state.ts` essentially saying the initialization and updating logic are in the component itself. I think it'll help in the future if there are any refactors if someone for whatever reason thinks `pageFilters` is not being used

logeekal replied:
Will do. Thanks.

michaelolo24 commented on x-pack/plugins/security_solution/public/common/components/filter_group/index.tsx:
```diff
@@ -0,0 +1,315 @@
+        chainingSystem,
+      };
+
+      // we want minimum of 4 fields to be dispayed in that particular order
```
❤️

logeekal commented on x-pack/plugins/security_solution/public/detections/components/detection_page_filters/index.tsx:
```diff
@@ -0,0 +1,55 @@
+};
+
+const arePropsEqual = (prevProps: FilterItemSetProps, newProps: FilterItemSetProps) => {
+  const _isEqual = isEqual(prevProps, newProps);
```
@michaelolo24 Yes, props were deep object and because of that it was rendering more than needed. But I do not remeber which prop exactly was creating the issue but I will check and update this comment again.

michaelolo24 commented:
As an aside I wonder if it makes sense to either hide these fields or potentially check ahead of time if the fields exist in the mappings before showing the options for them?

<img width="1457" alt="image" src="https://user-images.githubusercontent.com/17211684/207359767-3064eaaf-ebb2-42d3-b9b7-ac1b36623229.png">

logeekal committed:
Filter group : review feedback

logeekal commented:
> ed in design.
> @logeekal Is this one of variants of this `Filters component`, that we're reusing? Or is it our option?
> Also a question, is it an edge case than a common?

Thanks for checking it out @michaelolo24 . I agree with @codearos that, if we hide it, use will never know that something is wrong. 


@codearos , to answer your question, yes it is stock option from plugin when fields are not found. It is edge case when user has not setup any rules in elastic.

Do you think we can show any other error component to highlight user that to get rid of this error, they will have setup new rules?

PhilippeOberti commented:
@logeekal hey I just pulled down the branch and I'm getting this error
![Screen Shot 2022-12-15 at 11 22 29 AM](https://user-images.githubusercontent.com/17276605/207834793-6a2aa176-1e0b-4cd9-821d-0615a4a918f8.png)

Anything specific that I need to do to get setup?

logeekal commented:
> @logeekal hey I just pulled down the branch and I'm getting this error ![Screen Shot 2022-12-15 at 11 22 29 AM](https://user-images.githubusercontent.com/17276605/207834793-6a2aa176-1e0b-4cd9-821d-0615a4a918f8.png)
> 
> Anything specific that I need to do to get setup?

Thanks for checking it out @PhilippeOberti , I broke yesterday evening with my latest changes.. I am in process of updating and will let you know when it is done.

logeekal committed:
Merge main 👉🏽 current branch

logeekal committed:
fix: 🛠 tests + rule details page

logeekal committed:
fix type errors

logeekal committed:
refactor: change sort, title, exclude configs

logeekal committed:
fix one test

MadameSheema commented on x-pack/plugins/security_solution/cypress/e2e/detection_alerts/changing_alert_status.cy.ts:
```diff
@@ -194,6 +197,7 @@ describe('Changing alert status', () => {
           cy.get(SELECTED_ALERTS).should('have.text', `Selected ${numberOfAlertsToBeOpened} alert`);
 
           openAlerts();
+          cy.wait(10000);
```
Hardcoded waits are not allowed, please find another solution or ask for support on this.

logeekal replied:
Thanks for pointing it out.. it is a mistake.. I will remove this.

MadameSheema commented on x-pack/plugins/security_solution/cypress/tasks/alerts.ts:
```diff
@@ -108,8 +112,13 @@ export const closeAlerts = () => {
+  // which is preventing the popup to open upon button click below.
+  // Currently not sure what
+  waitForAlerts();
+  cy.wait(500);
```
Hardcoded waits are not allowed, please find another solution or ask for support on this.

logeekal replied:
🔨 Fixed. Please check the amendments here

MadameSheema replied:
Lots of thanks for removing all the waits ❤️

MadameSheema commented on x-pack/plugins/security_solution/cypress/e2e/detection_alerts/detection_page_filters.cy.ts:
```diff
@@ -0,0 +1,203 @@
+  });
+};
+
+const resetFilters = () => {
```
Why this is not extracted to a reusable method?

logeekal replied:
@MadameSheema .. corrected ✅

logeekal committed:
refactor: PR feedback. details below

1. Added spaces in filtergroup local storage key so that preferences are
   not shared across spaces.
2. Remove hardcoded wait and replaced with pipe function.
3. Removed old controls for Changing alert status

logeekal committed:
Corrected test to wait for all loading states

michaelolo24 commented on x-pack/plugins/security_solution/public/detections/pages/detection_engine/detection_engine.tsx:
```diff
@@ -324,33 +307,24 @@ const DetectionEnginePageComponent: React.FC<DetectionEngineComponentProps> = ({
-              </EuiFlexGroup>
-              <EuiSpacer size="m" />
+              <DetectionPageFilterSet
+                dataViewId={dataViewId as string}
```
Think it's worth having a placeholder and removing the cast? `dataViewId ? <DetectionPageFilterset ... /> : 'A dataview must be selected to use page filters'`

logeekal replied:
@michaelolo24 , done. No casting now.. The Filter group plugin takes care of invalid dataviewId and displaying the error message.

michaelolo24 replied:
👍🏾 ❤️

michaelolo24 commented on x-pack/plugins/security_solution/public/detection_engine/rule_details_ui/pages/rule_details/index.tsx:
```diff
@@ -624,6 +604,11 @@ const RuleDetailsPageComponent: React.FC<DetectionEngineComponentProps> = ({
   );
 
+  const isAlertTableLoading = useMemo(
+    () => loading || isAlertsLoading || !Array.isArray(detectionPageFilters),
```
There isn't a scenario where `detectionPageFilters` is somehow set to null or undefined at runtime causing the table to be stuck on a loading state? I didn't see any, but just wanted to confirm

logeekal replied:
@michaelolo24 , you are right. They are either empty array or array of Filters.

michaelolo24 commented:
Hey, thanks for making the changes from the last review. When I pulled down to test I saw one weird behavior, where clicking on a filter in the rule details page caused the page to jump which is either tied to some scroll behavior on that page or the popovers from the inputs:

https://user-images.githubusercontent.com/17211684/208451852-5f57976b-67d2-411d-a8d3-72ff96fad888.mov

logeekal commented:
> Hey, thanks for making the changes from the last review. When I pulled down to test I saw one weird behavior, where clicking on a filter in the rule details page caused the page to jump which is either tied to some scroll behavior on that page or the popovers from the inputs:
> Screen.Recording.2022-12-19.at.9.39.21.AM.mov

@michaelolo24 , this seems to be strange one but root of the issue lies with EuiPopover. I have raised [this issue](https://github.com/elastic/eui/issues/6494) with EUI Team. Although, I will keep looking into it and raise a PR with EUI if I am able to resolve it. 

Do you think we can move forward with the PR with this bug to be handled in Build candidates.

@MadameSheema ,  can we add this test in the test plan to make sure that before we release, this issue is fixed

logeekal committed:
fix: localstorage across space + no type cast

logeekal committed:
Merging main ➡️  current branch

logeekal commented on x-pack/plugins/security_solution/public/detection_engine/rule_details_ui/pages/rule_details/index.tsx:
```diff
@@ -817,19 +802,20 @@ const RuleDetailsPageComponent: React.FC<DetectionEngineComponentProps> = ({
-                          onFilterGroupChanged={onFilterGroupChangedCallback}
+                      <EuiFlexItem grow={true}>
+                        <DetectionPageFilterSet
+                          dataViewId={dataViewId as string}
```
@michaelolo24 : Done as well.

michaelolo24 commented:
> > Hey, thanks for making the changes from the last review. When I pulled down to test I saw one weird behavior, where clicking on a filter in the rule details page caused the page to jump which is either tied to some scroll behavior on that page or the popovers from the inputs:
> > Screen.Recording.2022-12-19.at.9.39.21.AM.mov
> 
> @michaelolo24 , this seems to be strange one but root of the issue lies with EuiPopover. I have raised [this issue](https://github.com/elastic/eui/issues/6494) with EUI Team. Although, I will keep looking into it and raise a PR with EUI if I am able to resolve it.
> 
> Do you think we can move forward with the PR with this bug to be handled in Build candidates.
> 
> @MadameSheema , can we add this test in the test plan to make sure that before we release, this issue is fixed

I worry about knowingly introducing bugs, but it definitely looks like an eui bug. I would say if we can a commitment from them to get eyes on that issue for 8.7 then you should be okay. In the event EUI cannot take a look or it can't be fixed this release, how difficult would it be to revert to use the `open,acknowledged,closed` buttons?

logeekal commented:
> I worry about knowingly introducing bugs, but it definitely looks like an eui bug. I would say if we can a commitment from them to get eyes on that issue for 8.7 then you should be okay. In the event EUI cannot take a look or it can't be fixed this release, how difficult would it be to revert to use the open,acknowledged,closed buttons?

I guess, we won't need to stop the release. If I remove ( i mean in the kibana presentation plugin that we are using) autofocus from the search box in the popover, the error goes away. It only occurs when autofocus is `true`.

So I guess we have a workaround.

michaelolo24 commented:
> > I worry about knowingly introducing bugs, but it definitely looks like an eui bug. I would say if we can a commitment from them to get eyes on that issue for 8.7 then you should be okay. In the event EUI cannot take a look or it can't be fixed this release, how difficult would it be to revert to use the open,acknowledged,closed buttons?
> 
> I guess, we won't need to stop the feature. If I remove ( i mean in the kibana presentation plugin that we are using) autofocus from the search box in the popover, the error goes away. It only occurs when autofocus is `true`.
> 
> So I guess we have a workaround.

Awesome! Yea, when you disable autofocus, I'll pull it down after to test it again and we should be good to go 😄 .

michaelolo24 commented:
For posterity, we spoke offline and decided given the time afforded for FF we can go ahead and merge this PR, since the EUI bug is non-breaking, and get a fix in on a separate branch so the work here does not go stale.

michaelolo24 reviewed (approved):
Thanks for doing this work! LGTM

logeekal committed:
Revert Rule Details Page

michaelolo24 commented on x-pack/plugins/security_solution/cypress/tasks/alerts.ts:
```diff
@@ -107,8 +116,32 @@ export const closeAlerts = () => {
+
+  /*
+   *
+   * Sometimes it takes some time for UI to attach event listener to
```
Thanks for the comment! 😄

michaelolo24 reviewed (approved):
Reviewed the changes, looks good! Thanks again

codearos reviewed (approved):
LGTM! Put Design QA comments here https://github.com/elastic/security-team/issues/4128#issuecomment-1346527838

banderror reviewed (approved):
LGTM. I couldn't find any changes directly related to the Rules area or D&R, although GH thinks we own 17 files in this PR:

<img width="333" alt="Screenshot 2022-12-23 at 12 06 03" src="https://user-images.githubusercontent.com/7359339/209325434-1871d6c7-7e7a-4ab6-9976-89ba683b4d2b.png">

Good old false positive ping by our CODEOWNERS file 😅 cc @YulNaumenko

logeekal committed:
Merge branch 'main' into feat/detection_page_filters

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/ts_project_linter --fix'

logeekal committed:
Merge branch 'main' into feat/detection_page_filters

logeekal committed:
Merge branch 'main' into feat/detection_page_filters

logeekal committed:
Merge branch 'main' into feat/detection_page_filters

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/notice'

logeekal committed:
Merge branch 'main' into feat/detection_page_filters

machadoum commented on x-pack/plugins/security_solution/public/common/components/filter_group/hooks/use_filter_update_to_url_sync.ts:
```diff
@@ -0,0 +1,53 @@
+
+  useEffect(() => {
+    if (!formattedFilters) return;
+    dispatch(
```
I am a bit late here... But could you use the `useUpdateUrlParam` hook instead of directly calling dispatch?
